File: ios/iLoppis/UI/Screens/Cashier/CashierHeartbeat.swift
import Foundation

struct CashierPresenceSnapshot: Equatable {
    let clientState: CashierClientState
    let pendingPurchasesCount: Int
    let displayName: String?
}

extension CashierState {
    func heartbeatSnapshot() -> CashierPresenceSnapshot {
        let pendingPurchasesCount = transactions.isEmpty ? 0 : 1
        let clientState: CashierClientState

        if isProcessingPayment {
            clientState = .submitting
        } else if pendingPurchasesCount > 0 {
            clientState = .activeTransaction
        } else {
            clientState = .idle
        }

        return CashierPresenceSnapshot(
            clientState: clientState,
            pendingPurchasesCount: pendingPurchasesCount,
            displayName: heartbeatDisplayName
        )
    }
}

final class CashierHeartbeatCoordinator {
    typealias RequestFactory = @Sendable () async -> CashierPresenceHeartbeatRequest?
    typealias HeartbeatSender = @Sendable (CashierPresenceHeartbeatRequest) async throws -> CashierPresenceHeartbeatResponse
    typealias ResponseHandler = @Sendable (CashierPresenceHeartbeatResponse) async -> Void
    typealias ErrorHandler = @Sendable (Error) async -> Void

    private let intervalNanoseconds: UInt64
    private let shouldRun: () -> Bool
    private let requestFactory: RequestFactory
    private let sendHeartbeat: HeartbeatSender
    private let onHeartbeatResponse: ResponseHandler
    private let onHeartbeatFailure: ErrorHandler
    private let sessionManager: RegisterSessionManager?

    private var task: Task<Void, Never>?

    init(
        intervalNanoseconds: UInt64 = 15_000_000_000,
        shouldRun: @escaping () -> Bool,
        requestFactory: @escaping RequestFactory,
        sendHeartbeat: @escaping HeartbeatSender,
        onHeartbeatResponse: @escaping ResponseHandler,
        onHeartbeatFailure: @escaping ErrorHandler,
        sessionManager: RegisterSessionManager? = nil
    ) {
        self.intervalNanoseconds = intervalNanoseconds
        self.shouldRun = shouldRun
        self.requestFactory = requestFactory
        self.sendHeartbeat = sendHeartbeat
        self.onHeartbeatResponse = onHeartbeatResponse
        self.onHeartbeatFailure = onHeartbeatFailure
        self.sessionManager = sessionManager
    }

    deinit {
        task?.cancel()
    }

    func start() {
        guard task == nil, shouldRun() else {
            return
        }

        let shouldRun = self.shouldRun
        let requestFactory = self.requestFactory
        let sendHeartbeat = self.sendHeartbeat
        let onHeartbeatResponse = self.onHeartbeatResponse
        let onHeartbeatFailure = self.onHeartbeatFailure
        let intervalNanoseconds = self.intervalNanoseconds
        let sessionManager = self.sessionManager

        task = Task { [weak self] in
            defer { self?.task = nil }

            while !Task.isCancelled {
                guard shouldRun(), let request = await requestFactory() else {
                    break
                }

                do {
                    let response = try await sendHeartbeat(request)
                    await MainActor.run {
                        sessionManager?.clearPendingLifecycleEvent(
                            expectedLifecycleEvent: request.lifecycleEventType,
                            expectedSessionId: request.sessionId
                        )
                        if request.lifecycleEventType == nil {
                            sessionManager?.recordSync()
                        }
                    }
                    await onHeartbeatResponse(response)
                } catch is CancellationError {
                    break
                } catch {
                    await onHeartbeatFailure(error)
                }

                do {
                    try await Task.sleep(nanoseconds: intervalNanoseconds)
                } catch {
                    break
                }
            }
        }
    }

    func stop() {
        task?.cancel()
        task = nil
    }
}

File: android/app/src/main/java/se/iloppis/app/ui/screens/cashier/CashierHeartbeat.kt
package se.iloppis.app.ui.screens.cashier

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import se.iloppis.app.data.RegisterSessionManager
import se.iloppis.app.network.cashier.CashierClientState
import se.iloppis.app.network.cashier.CashierPresenceHeartbeatRequest
import se.iloppis.app.network.cashier.CashierPresenceHeartbeatResponse

internal data class CashierPresenceSnapshot(
    val clientState: CashierClientState,
    val pendingPurchasesCount: Int,
    val displayName: String?
)

internal fun CashierUiState.toCashierPresenceSnapshot(rawPendingPurchasesCount: Int): CashierPresenceSnapshot {
    val activeTransactionCount = if (transactions.isNotEmpty()) 1 else 0
    val pendingPurchasesCount = rawPendingPurchasesCount + activeTransactionCount
    val clientState = when {
        isProcessingPayment -> CashierClientState.CASHIER_CLIENT_STATE_SUBMITTING
        pendingPurchasesCount > 0 -> CashierClientState.CASHIER_CLIENT_STATE_ACTIVE_TRANSACTION
        else -> CashierClientState.CASHIER_CLIENT_STATE_IDLE
    }
    return CashierPresenceSnapshot(
        clientState = clientState,
        pendingPurchasesCount = pendingPurchasesCount,
        displayName = heartbeatDisplayName
    )
}

/** Builds a heartbeat request without serializing the session manager itself. */
internal fun CashierPresenceSnapshot.toHeartbeatRequest(
    clientType: se.iloppis.app.network.cashier.CashierClientType,
    sessionManager: RegisterSessionManager
): CashierPresenceHeartbeatRequest {
    val session = sessionManager.getCurrent()
    return CashierPresenceHeartbeatRequest(
        clientState = clientState,
        pendingPurchasesCount = pendingPurchasesCount,
        clientType = clientType,
        displayName = displayName,
        lifecycleEventType = session?.pendingLifecycleEvent,
        registerId = session?.registerId,
        sessionId = session?.sessionId
    )
}

internal class CashierHeartbeatCoordinator(
    private val scope: CoroutineScope,
    private val intervalMs: Long = 15_000L,
    private val shouldRun: () -> Boolean,
    private val requestFactory: () -> CashierPresenceHeartbeatRequest,
    private val sendHeartbeat: suspend (CashierPresenceHeartbeatRequest) -> CashierPresenceHeartbeatResponse,
    private val onHeartbeatResponse: (CashierPresenceHeartbeatResponse) -> Unit,
    private val onHeartbeatFailure: (Throwable) -> Unit,
    /** Clears acknowledged lifecycle transitions after successful heartbeats. */
    private val sessionManager: RegisterSessionManager? = null
) {
    private var job: Job? = null

    /** Starts one heartbeat loop when [shouldRun] permits it; repeated calls are idempotent. */
    fun start() {
        if (job?.isActive == true || !shouldRun()) {
            return
        }

        job = scope.launch {
            while (isActive) {
                if (!shouldRun()) {
                    break
                }

                try {
                    val request = requestFactory()
                    val response = sendHeartbeat(request)
                    sessionManager?.clearPendingLifecycleEvent(
                        expectedLifecycleEvent = request.lifecycleEventType,
                        expectedSessionId = request.sessionId
                    )
                    if (request.lifecycleEventType == null) {
                        sessionManager?.recordSync()
                    }
                    onHeartbeatResponse(response)
                } catch (cancellationException: CancellationException) {
                    throw cancellationException
                } catch (t: Throwable) {
                    onHeartbeatFailure(t)
                }

                delay(intervalMs)
            }
        }
    }

    /** Cancels the active heartbeat loop; repeated calls are safe. */
    fun stop() {
        job?.cancel()
        job = null
    }
}
